File: abl/src/abl/wmes/BehaviorStatusWME.java
package abl.wmes;

import org.json.simple.JSONObject;
import wm.WME;

public class BehaviorStatusWME extends WME {
    /** Whether or not this WME is on an active behavior tree */
    private boolean onTree = false;

    /** ID of the character this status pertains to */
    private int characterID;

    /** Name of the behavior that completed or was terminated */
    private String behaviorName;

    /** Status of the behavior (e.g., "Success", "Failure", "Interrupted") */
    private String status;

    /**
     * Constructor that populates the WME fields from a JSONObject.
     * This JSONObject is expected to be parsed from the JSON data
     * sent by the Unity C# BehaviorStatusWME.
     *
     * @param data The JSONObject containing the WME data.
     *             Expected keys: "characterID", "behaviorName", "status".
     */
    public BehaviorStatusWME(JSONObject data) {
        this.onTree = false;
        this.characterID = (int) (long) data.get("characterID");
        this.behaviorName = (String) data.get("behaviorName");
        this.status = (String) data.get("status");
    }

    /**
     * Provides a string representation of this WME, useful for logging and
     * debugging.
     *
     * @return A string detailing the WME's fields.
     */
    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();
        builder.append("BehaviorStatusWME: \n")
                .append("  CharacterID: ").append(this.characterID).append("\n")
                .append("  BehaviorName: ").append(this.behaviorName).append("\n")
                .append("  Status: ").append(this.status);
        return builder.toString();
    }

    // Getter methods for the WME's fields

    public int getCharacterID() {
        return characterID;
    }

    public String getBehaviorName() {
        return behaviorName;
    }

    public String getStatus() {
        return status;
    }

    public boolean isOnTree() {
        return onTree;
    }
}

File: oops-all-bards/Assets/Scripts/ABL/WMEs/BehaviorStatusWME.cs
using UnityEngine;

/// <summary>
/// Represents a Working Memory Element (WME) that signals the status 
/// of a completed or terminated ICharacterBehavior in Unity.
/// This is intended to be created in Unity and then packaged into an ABLMessage
/// to be sent to and processed by the ABL system.
/// </summary>
[System.Serializable]
public class BehaviorStatusWME
{
    [SerializeField] private int characterID;
    [SerializeField] private string behaviorName; // e.g., "Move", "Observe", "CalmConfrontation"
    [SerializeField] private string status;       // e.g., "Success", "Failure", "Interrupted"

    private const int MESSAGE_CODE = 3;
    private const string MESSAGE_TYPE = "BehaviorStatusWME";

    /// <summary>
    /// Constructor for BehaviorStatusWME.
    /// </summary>
    /// <param name="charID">The ID of the character whose behavior status is being reported.</param>
    /// <param name="behName">The name of the behavior (typically from ICharacterBehavior.GetBehaviorName()).</param>
    /// <param name="stat">The completion status: "Success", "Failure", or "Interrupted".</param>
    public BehaviorStatusWME(int charID, string behName, string stat)
    {
        this.characterID = charID;
        this.behaviorName = behName;
        this.status = stat;
    }

    // Public properties to access the fields
    public int CharacterID
    {
        get { return this.characterID; }
    }

    public string BehaviorName
    {
        get { return this.behaviorName; }
    }

    public string Status
    {
        get { return this.status; }
    }

    /// <summary>
    /// Packages this WME into an ABLMessage for sending to the ABL server.
    /// The server uses the message code (3) to identify how to parse the JSON data.
    /// </summary>
    /// <returns>An ABLMessage containing this WME's data.</returns>
    public ABLMessage ToABLMessage()
    {
        string jsonData = JsonUtility.ToJson(this);
        ABLMessage message = new ABLMessage(MESSAGE_CODE, MESSAGE_TYPE, jsonData);
        return message;
    }

    /// <summary>
    /// Utility method to convert this WME directly to a JSON string for logging or other purposes.
    /// </summary>
    /// <returns>A JSON string representation of this object.</returns>
    public string ToJson()
    {
        return JsonUtility.ToJson(this);
    }
}
